### DISCUSSION TITLE
importorskip usage with "from" imports?

### DISCUSSION THREAD

**Discussion by wimglenn** [pytest-dev/pytest#12894] [Ideas] [answered]
I like that I can do things like this in a simple test, rather than use some janky setup code for ensuring that a runtime version and metadata version are single-sourced at build-time:

```
import pytest
import mypkg

def test_metadata_version_in_sync_with_runtime_version():
    importlib_metadata = pytest.importorskip("importlib.metadata")
    assert importlib_metadata.version("mypkg") == mypkg.__version__
```

`importlib.metadata` was new in 3.8, and this isn't a test that really needs to run on all Python versions. Just one runtime should be sufficient to catch the basic issue.

However, it relies on the fact that `importlib.metadata.version` arrived at the same time as `importlib.metadata`. Is there a similar usage available when the _module_ is available, but there is a new name in the module namespace?

An example could be [`importlib.medata.packages_distributions`](https://docs.python.org/3/library/importlib.metadata.html#importlib.metadata.packages_distributions) which was added in 3.10.

Something like this:

```
def test_foo():
    try:
        from importlib.metadata import packages_distributions
    except ImportError:
        pytest.skip(...)
    ...
```

But, obviously, using the `importorskip` functionality instead of a try/except.

I just use `importlib.metadata` as a motivating example, of course the question goes for any module / module namespace.

If this is something that could be worth adding to `pytest.importorskip`, can we discuss what a good API would look like?

**Comment by The-Compiler** [comment, marked as answer]
You could do:

```python
import importlib.metadata
import pytest

@pytest.mark.skipif(not hasattr(importlib.metadata, "packages_distributions"), reason="Needs packages_distributions")
def test_foo():
    ... = importlib.metadata.packages_distributions()
```

and possibly even move that into a small helper if you want to have a nice `reason` without the duplication. That would have the additional benefit that the test gets skipped at collection time without having to first run it (and possibly set up fixtures) before finding out it's actually skipped.
